File: web/src/color.js
// Color math for the topowall web UI. Mirrors crates/render/src/color.rs so
// colors picked in the browser mean exactly the same thing to the renderer.
//
// RGBA objects: { r, g, b } in 0..255 (floats allowed), a in 0..1.
// HSV objects:  { h } degrees 0..360, { s, v } in 0..1.

export const clamp = (v, lo, hi) => Math.min(hi, Math.max(lo, v));

// ── Hex ─────────────────────────────────────────────────────────────────────

/** Parse #rgb, #rgba, #rrggbb or #rrggbbaa (the # is optional). Returns null if invalid. */
export function hexToRgba(input) {
  const hex = String(input).trim().replace(/^#/, "");
  if (!/^[0-9a-f]+$/i.test(hex)) return null;
  let r, g, b, a = 255;
  if (hex.length === 3 || hex.length === 4) {
    [r, g, b] = [0, 1, 2].map((i) => parseInt(hex[i] + hex[i], 16));
    if (hex.length === 4) a = parseInt(hex[3] + hex[3], 16);
  } else if (hex.length === 6 || hex.length === 8) {
    [r, g, b] = [0, 2, 4].map((i) => parseInt(hex.slice(i, i + 2), 16));
    if (hex.length === 8) a = parseInt(hex.slice(6, 8), 16);
  } else {
    return null;
  }
  return { r, g, b, a: a / 255 };
}

const byteHex = (v) => Math.round(clamp(v, 0, 255)).toString(16).padStart(2, "0");

/** Format as #rrggbb, or #rrggbbaa when alpha < 1. */
export function rgbaToHex({ r, g, b, a = 1 }) {
  const alpha = Math.round(clamp(a, 0, 1) * 255);
  return "#" + byteHex(r) + byteHex(g) + byteHex(b) + (alpha < 255 ? byteHex(alpha) : "");
}

// ── HSV ─────────────────────────────────────────────────────────────────────

/** RGB → HSV. Hue is NaN for grays and saturation is NaN for black, so callers can keep their previous values. */
export function rgbToHsv({ r, g, b }) {
  const [rn, gn, bn] = [r / 255, g / 255, b / 255];
  const max = Math.max(rn, gn, bn);
  const min = Math.min(rn, gn, bn);
  const d = max - min;
  let h = NaN;
  if (d > 1e-9) {
    if (max === rn) h = ((gn - bn) / d) % 6;
    else if (max === gn) h = (bn - rn) / d + 2;
    else h = (rn - gn) / d + 4;
    h = (h * 60 + 360) % 360;
  }
  const s = max > 1e-9 ? d / max : NaN;
  return { h, s, v: max };
}

export function hsvToRgb({ h, s, v }) {
  const f = (n) => {
    const k = (n + h / 60) % 6;
    return v - v * s * Math.max(0, Math.min(k, 4 - k, 1));
  };
  return { r: f(5) * 255, g: f(3) * 255, b: f(1) * 255 };
}

// ── OKLab / OKLCH ───────────────────────────────────────────────────────────

const toLinear = (c) => (c <= 0.04045 ? c / 12.92 : ((c + 0.055) / 1.055) ** 2.4);
const toSrgb = (c) => (c <= 0.0031308 ? 12.92 * c : 1.055 * c ** (1 / 2.4) - 0.055);

function linearFromOklab(L, a, b) {
  const l = (L + 0.3963377774 * a + 0.2158037573 * b) ** 3;
  const m = (L - 0.1055613458 * a - 0.0638541728 * b) ** 3;
  const s = (L - 0.0894841775 * a - 1.291485548 * b) ** 3;
  return [
    4.0767416621 * l - 3.3077115913 * m + 0.2309699292 * s,
    -1.2684380046 * l + 2.6097574011 * m - 0.3413193965 * s,
    -0.0041960863 * l - 0.7034186147 * m + 1.707614701 * s,
  ];
}

export function rgbToOklch({ r, g, b }) {
  const [lr, lg, lb] = [r, g, b].map((c) => toLinear(c / 255));
  const l = Math.cbrt(0.4122214708 * lr + 0.5363325363 * lg + 0.0514459929 * lb);
  const m = Math.cbrt(0.2119034982 * lr + 0.6806995451 * lg + 0.1073969566 * lb);
  const s = Math.cbrt(0.0883024619 * lr + 0.2817188376 * lg + 0.6299787005 * lb);
  const L = 0.2104542553 * l + 0.793617785 * m - 0.0040720468 * s;
  const A = 1.9779984951 * l - 2.428592205 * m + 0.4505937099 * s;
  const B = 0.0259040371 * l + 0.7827717662 * m - 0.808675766 * s;
  return { l: L, c: Math.hypot(A, B), h: ((Math.atan2(B, A) * 180) / Math.PI + 360) % 360 };
}

/** OKLCH → RGB, reducing chroma until the color fits sRGB (same as the renderer). */
export function oklchToRgb({ l, c, h }) {
  const rad = (h * Math.PI) / 180;
  let chroma = c;
  let lin;
  for (let i = 0; i < 24; i++) {
    lin = linearFromOklab(l, chroma * Math.cos(rad), chroma * Math.sin(rad));
    if (lin.every((v) => v >= -1e-4 && v <= 1.0001)) break;
    chroma *= 0.9;
  }
  const [r, g, b] = lin.map((v) => toSrgb(clamp(v, 0, 1)) * 255);
  return { r, g, b };
}

export function formatOklch({ l, c, h }) {
  return `oklch(${l.toFixed(3)} ${c.toFixed(3)} ${h.toFixed(1)})`;
}

// ── Parsing any supported string ────────────────────────────────────────────

/** Parse hex, oklch(L C H [/ A]) or rgb()/rgba(). Returns RGBA or null. */
export function parseColor(input) {
  const s = String(input).trim().toLowerCase();
  const fn = s.match(/^(oklch|rgba?)\((.*)\)$/);
  if (!fn) return hexToRgba(s);
  const parts = fn[2].split(/[\s,/]+/).filter(Boolean);
  const num = (p, scale = 1) => (p.endsWith("%") ? (parseFloat(p) / 100) * scale : parseFloat(p));
  if (parts.length < 3 || parts.slice(0, 3).some((p) => Number.isNaN(parseFloat(p)))) return null;
  const alpha = parts[3] !== undefined ? clamp(num(parts[3]), 0, 1) : 1;
  if (fn[1] === "oklch") {
    const rgb = oklchToRgb({ l: num(parts[0]), c: num(parts[1], 0.4), h: parseFloat(parts[2]) });
    return { ...rgb, a: alpha };
  }
  const [r, g, b] = parts.slice(0, 3).map((p) => clamp(num(p, 255), 0, 255));
  return { r, g, b, a: alpha };
}

File: crates/kit/src/color.rs
//! Colors: parsing (`#rgb`, `#rrggbb`, `#rrggbbaa`, `oklch(L C H)`) and OKLab/OKLCH math.

use anyhow::{bail, Context, Result};

/// sRGB-encoded color with straight alpha, components in 0..=1.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Color {
    pub r: f32,
    pub g: f32,
    pub b: f32,
    pub a: f32,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Oklch {
    pub l: f32,
    pub c: f32,
    /// Hue in degrees.
    pub h: f32,
}

fn to_linear(c: f32) -> f32 {
    if c <= 0.04045 {
        c / 12.92
    } else {
        ((c + 0.055) / 1.055).powf(2.4)
    }
}

fn to_srgb(c: f32) -> f32 {
    let c = c.clamp(0.0, 1.0);
    if c <= 0.003_130_8 {
        12.92 * c
    } else {
        1.055 * c.powf(1.0 / 2.4) - 0.055
    }
}

impl Color {
    pub const BLACK: Color = Color {
        r: 0.0,
        g: 0.0,
        b: 0.0,
        a: 1.0,
    };

    pub fn rgb8(r: u8, g: u8, b: u8) -> Self {
        Self {
            r: r as f32 / 255.0,
            g: g as f32 / 255.0,
            b: b as f32 / 255.0,
            a: 1.0,
        }
    }

    pub fn parse(s: &str) -> Result<Self> {
        let t = s.trim();
        if let Some(inner) = t.strip_prefix("oklch(").and_then(|x| x.strip_suffix(')')) {
            let parts: Vec<&str> = inner
                .split(|c: char| c.is_whitespace() || c == ',' || c == '/')
                .filter(|p| !p.is_empty())
                .collect();
            if parts.len() < 3 {
                bail!("oklch() needs L C H, got '{s}'");
            }
            let num = |p: &str| -> Result<f32> {
                match p.strip_suffix('%') {
                    Some(pct) => Ok(pct.parse::<f32>()? / 100.0),
                    None => Ok(p.trim_end_matches("deg").parse::<f32>()?),
                }
            };
            let mut c = Oklch {
                l: num(parts[0])?,
                c: num(parts[1])?,
                h: num(parts[2])?,
            }
            .to_color();
            if let Some(a) = parts.get(3) {
                c.a = num(a)?;
            }
            return Ok(c);
        }
        let hex = t.strip_prefix('#').unwrap_or(t);
        let digits: Vec<u8> = hex
            .chars()
            .map(|c| c.to_digit(16).map(|d| d as u8))
            .collect::<Option<_>>()
            .with_context(|| format!("invalid color '{s}' (use #rrggbb or oklch(L C H))"))?;
        let byte = |i: usize| digits[i] * 16 + digits[i + 1];
        let (r, g, b, a) = match digits.len() {
            3 => (digits[0] * 17, digits[1] * 17, digits[2] * 17, 255),
            6 => (byte(0), byte(2), byte(4), 255),
            8 => (byte(0), byte(2), byte(4), byte(6)),
            _ => bail!("invalid color '{s}' (use #rrggbb or oklch(L C H))"),
        };
        Ok(Self {
            r: r as f32 / 255.0,
            g: g as f32 / 255.0,
            b: b as f32 / 255.0,
            a: a as f32 / 255.0,
        })
    }

    pub fn to_hex(self) -> String {
        let q = |v: f32| (v.clamp(0.0, 1.0) * 255.0).round() as u8;
        if q(self.a) == 255 {
            format!("#{:02x}{:02x}{:02x}", q(self.r), q(self.g), q(self.b))
        } else {
            format!(
                "#{:02x}{:02x}{:02x}{:02x}",
                q(self.r),
                q(self.g),
                q(self.b),
                q(self.a)
            )
        }
    }

    pub fn to_array(self) -> [f32; 4] {
        [self.r, self.g, self.b, self.a]
    }

    pub fn to_oklab(self) -> [f32; 3] {
        let (r, g, b) = (to_linear(self.r), to_linear(self.g), to_linear(self.b));
        let l = (0.412_221_46 * r + 0.536_332_55 * g + 0.051_445_995 * b).cbrt();
        let m = (0.211_903_5 * r + 0.680_699_5 * g + 0.107_396_96 * b).cbrt();
        let s = (0.088_302_46 * r + 0.281_718_85 * g + 0.629_978_7 * b).cbrt();
        [
            0.210_454_26 * l + 0.793_617_8 * m - 0.004_072_047 * s,
            1.977_998_5 * l - 2.428_592_2 * m + 0.450_593_7 * s,
            0.025_904_037 * l + 0.782_771_77 * m - 0.808_675_77 * s,
        ]
    }

    pub fn from_oklab([ll, a, b]: [f32; 3]) -> Self {
        let l = (ll + 0.396_337_78 * a + 0.215_803_76 * b).powi(3);
        let m = (ll - 0.105_561_346 * a - 0.063_854_17 * b).powi(3);
        let s = (ll - 0.089_484_18 * a - 1.291_485_5 * b).powi(3);
        Self {
            r: to_srgb(4.076_741_7 * l - 3.307_711_6 * m + 0.230_969_94 * s),
            g: to_srgb(-1.268_438 * l + 2.609_757_4 * m - 0.341_319_38 * s),
            b: to_srgb(-0.004_196_086_3 * l - 0.703_418_6 * m + 1.707_614_7 * s),
            a: 1.0,
        }
    }

    pub fn to_oklch(self) -> Oklch {
        let [l, a, b] = self.to_oklab();
        Oklch {
            l,
            c: a.hypot(b),
            h: b.atan2(a).to_degrees().rem_euclid(360.0),
        }
    }

    /// Perceptual mix in OKLab (t = 0 → self, 1 → other).
    pub fn mix(self, other: Color, t: f32) -> Color {
        let (x, y) = (self.to_oklab(), other.to_oklab());
        let mut c = Color::from_oklab([
            x[0] + (y[0] - x[0]) * t,
            x[1] + (y[1] - x[1]) * t,
            x[2] + (y[2] - x[2]) * t,
        ]);
        c.a = self.a + (other.a - self.a) * t;
        c
    }
}

impl Oklch {
    /// Convert to sRGB, reducing chroma until the color fits the sRGB gamut.
    pub fn to_color(self) -> Color {
        let mut c = self.c;
        for _ in 0..24 {
            let (a, b) = (c * self.h.to_radians().cos(), c * self.h.to_radians().sin());
            let l = self.l;
            let lin = [
                (l + 0.396_337_78 * a + 0.215_803_76 * b).powi(3),
                (l - 0.105_561_346 * a - 0.063_854_17 * b).powi(3),
                (l - 0.089_484_18 * a - 1.291_485_5 * b).powi(3),
            ];
            let rgb = [
                4.076_741_7 * lin[0] - 3.307_711_6 * lin[1] + 0.230_969_94 * lin[2],
                -1.268_438 * lin[0] + 2.609_757_4 * lin[1] - 0.341_319_38 * lin[2],
                -0.004_196_086_3 * lin[0] - 0.703_418_6 * lin[1] + 1.707_614_7 * lin[2],
            ];
            if rgb.iter().all(|v| (-1e-4..=1.0001).contains(v)) {
                break;
            }
            c *= 0.9;
        }
        Color::from_oklab([
            self.l,
            c * self.h.to_radians().cos(),
            c * self.h.to_radians().sin(),
        ])
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hex_roundtrip() {
        for h in ["#000000", "#92aca0", "#3e5d58", "#00aeef", "#ffffff"] {
            assert_eq!(Color::parse(h).unwrap().to_hex(), h);
        }
        assert_eq!(Color::parse("#fff").unwrap().to_hex(), "#ffffff");
        assert_eq!(Color::parse("#11223380").unwrap().to_hex(), "#11223380");
    }

    #[test]
    fn oklch_roundtrip() {
        let c = Color::parse("#92aca0").unwrap();
        let back = c.to_oklch().to_color();
        assert_eq!(back.to_hex(), "#92aca0");
    }

    #[test]
    fn parses_oklch() {
        let c = Color::parse("oklch(0.72 0.05 234)").unwrap();
        let lch = c.to_oklch();
        assert!((lch.l - 0.72).abs() < 0.01 && (lch.h - 234.0).abs() < 1.0);
    }
}
